app/post_helpers.go

```diff
@@ -0,0 +1,193 @@
+	}
+
+	for {
+		distance := getDistance(accessible, inaccessible)
+		if distance == 1 {
+			break
+		}
+		guess := min(accessible, inaccessible) + distance/2
+		guessIsAccessible := posts[order[guess]].CreateAt >= earliestAccessibleTime
+		if guessIsAccessible {
+			accessible = guess
+		} else {
+			inaccessible = guess
+		}
+	}
+
+	return postAccessibleBounds{
+		accessible:   accessible,
+		inaccessible: inaccessible,
+	}
+}
+
+type filterPostOptions struct {
+	assumeSortedCreatedAt bool
+}
+
+// make no assumptions about ordering, go through posts one by one
+// this is the slower fallback that is still safe if we we can not
+// assume posts are ordered by CreatedAt
+func linearFilterPosts(postList *model.PostList, earliestAccessibleTime int64) {
+	// filter Posts
+	posts := postList.Posts
+
+	newPostOrder := []string{}
+	for _, postId := range postList.Order {
+		if posts[postId].CreateAt < earliestAccessibleTime {
+			postList.HasInaccessiblePosts = true
+			delete(posts, postId)
+		} else {
+			newPostOrder = append(newPostOrder, postId)
+		}
+	}
+
+	// it can happen that some post list results don't have all posts in the Order field.
+	// for example GetPosts in the CollapsedThreads = false path, parents are not added
+	// to Order
+	for postId, post := range posts {
+		if post.CreateAt < earliestAccessibleTime {
+			postList.HasInaccessiblePosts = true
+			delete(posts, postId)
+		}
+	}
+
+	postList.Order = newPostOrder
+}
+
+// filterInaccessiblePosts filters out the posts, past the cloud limit
+func (a *App) filterInaccessiblePosts(postList *model.PostList, options filterPostOptions) *model.AppError {
+	if postList == nil || postList.Posts == nil || len(postList.Posts) == 0 {
+		return nil
+	}
+
+	lastAccessiblePostTime, appErr := a.GetLastAccessiblePostTime()
+	if appErr != nil {
+		return model.NewAppError("filterInaccessiblePosts", "app.last_accessible_post.app_error", nil, appErr.Error(), http.StatusInternalServerError)
+	} else if lastAccessiblePostTime == 0 {
+		// No need to filter, all posts are accessible
+		return nil
+	}
+
+	if options.assumeSortedCreatedAt {
+		bounds := getTimeSortedPostAccessibleBounds(postList, lastAccessiblePostTime)
+		if bounds.allAccessible() {
+			return nil
+		}
+		if bounds.allInaccessible() {
+			postList.Posts = map[string]*model.Post{}
+			postList.Order = []string{}
+			if len(postList.Posts) > 0 {
+				postList.HasInaccessiblePosts = true
+			}
+			return nil
+		}
+		postList.HasInaccessiblePosts = true
+
+		var otherInaccessibleBound int
+		var otherAccessibleBound int
+		if bounds.accessible > bounds.inaccessible {
+			otherInaccessibleBound = 0
+			otherAccessibleBound = len(postList.Order) - 1
+		} else {
+			otherInaccessibleBound = len(postList.Order) - 1
+			otherAccessibleBound = 0
+		}
+		order := postList.Order
+		for i := min(bounds.inaccessible, otherInaccessibleBound); i <= maxInt(bounds.inaccessible, otherInaccessibleBound); i++ {
+			delete(postList.Posts, order[i])
```

Comment:
Could it be beneficial to decide the `delete/make` operation conditionally, in case the `len` of inaccessible posts is much higher than the accessible posts. In those cases, we can create a new map instead of deleting keys?